### PR TITLE
Fix exception handling and counter update

### PR DESCRIPTION
**Pull request by borislin**
Fixed exception handling when parsing each document and counter update

### CODE REVIEW

**Comment by lintool** [comment]
Hi @borislin I'm lacking the broader context - can you explain how the code was buggy before?

**Comment by lintool** [comment]
@borislin Hrm. Interesting that it doesn't have an impact on the regressions...

**Comment by borislin** [comment]
@lintool @Peilin-Yang I just pushed one more commit. This won't affect the effectiveness of our program for now. But it fixed an issue if user accidentally accessed the iterator again after closing it (marked by `atEOF` rather than the `NoSuchElementException`).

Review this change and I'll merge it after that...

**Comment by lintool** [comment]
lgtm, but let's see if @Peilin-Yang has any comments.

File: src/main/java/io/anserini/index/IndexCollection.java [hunk 1]
```diff
@@ -183,17 +183,24 @@ public void run() {
...
+          boolean hasNext = false;
           try {
-            d = iter.next();
+            hasNext = iter.hasNext();
           } catch (NoSuchElementException e1) {
-            continue;
-          } catch (Exception e2) { // TODO: update related counters (#317)
+            break;
+          } catch (Exception e2) {
             LOG.warn("Exception when parsing document: ", e2);
```
Comment (Peilin-Yang) [inline src/main/java/io/anserini/index/IndexCollection.java]: We are just reading the next document so it should be "Exception when reading document: ", right?

**Comment by borislin** [comment]
@Peilin-Yang The reason I catch `IOException` in `hasNext()` is because the overwritten method `hasNext()` in `java.util.Iterator` doesn't throw `IOException`. We don't have this problem if `Segment` doesn't extend to `Iterator`.

**Comment by Peilin-Yang** [comment]
Then we will have to pass the global error counter(s) to the collection class and update the error counter(s) when encountering IOException

**Comment by Peilin-Yang** [comment]
Or we can throw a RuntimeException when encountering the IOException

**Comment by borislin** [comment]
@Peilin-Yang When we encounter an `IOException`, we should throw an exception that's different than the one that's thrown when parsing a document (right now we use `RuntimeException`) so we can differentiate them in `IndexCollection` to determine whether to continue reading/parsing more documents or not.

How about changing current `RuntimeException` to `ParseException` and throw `RuntimeCollection` when we encounter `IOException` in `hasNext()`?

**Comment by Peilin-Yang** [comment]
@borislin sounds good to me

**Comment by lintool** [comment]
@borislin In general, I'm in favor of throwing our own exceptions.

**Comment by borislin** [comment]
I've decided to throw `RuntimeException` when parsing document as well as for `IOException`. So, different than the approach I've discussed earlier. Reason is that `ParseException` is a checked exception and must be handled locally or declared in the method signature, conflicting with the method signature for `hasNext()` as discussed earlier.

In short, we are now using the exception message (containing "IOException" or not) to determine whether the `RuntimeException` is intended to be an `IOException` or not.

**Comment by Peilin-Yang** [comment]
Yea, `ParseException` is not a subclass of `RuntimeException`...

Regarding the counters, I think it is better to have just three of them?

- Indexed
- Error (including parsing error, the bug of our code, etc)
- Empty

I doubt we really need that detailed counters. It is likely that nobody will remember what is exactly the rule of `skipped`....

**Comment by borislin** [comment]
@Peilin-Yang 

Then we will have a lot of errors since there are a lot of deleted tweets and I think deleted tweets are technically not errors.

Can we deal with this specific index counter issue in #317 and get this PR shipped first? Let's fix the buggy code first.

I've also re-run all regression tests excluding cw12 and they have all passed.

**Comment by Peilin-Yang** [comment]
Shouldn't we just fix it in this PR? I thought this is for #317, no?

**Comment by borislin** [comment]
There are more places that we need to fix the counter issue. I believe we need to get #353 done first then #317.

**Pull request merged**